openvdb_houdini/GEO_PrimVDB.cc

```diff
@@ -2604,62 +2860,190 @@ geoSetVDBStreamCompression(openvdb::io::Stream& vos, bool backwards_compatible)
 }
 
 bool
-GEO_PrimVDB::saveVDB(UT_JSONWriter &w) const
+GEO_PrimVDB::saveVDB(UT_JSONWriter &w, const GA_SaveMap &sm,
+		     bool as_shmem) const
 {
-    bool        ok = true;
+    bool	ok = true;
 
     try
     {
-        openvdb::GridCPtrVec grids;
-        grids.push_back(getConstGridPtr());
-
-        UT_JSONWriter::TiledStream os(w);
-
-        openvdb::io::Stream                     vos(os);
-        openvdb::MetaMap                        meta;
-
-        geoSetVDBStreamCompression(vos,
-            UT_EnvControl::getInt(ENV_HOUDINI13_VOLUME_COMPATIBILITY));
-
-        // Visual C++ requires a default meta object declared on the stack
-        vos.write(grids, meta);
+	openvdb::GridCPtrVec grids;
+	grids.push_back(getConstGridPtr());
+
+	if (as_shmem)
+	{
+	    openvdb::MetaMap meta;
+
+	    UT_String shmem_owner;
+	    sm.getOptions().importOption("geo:sharedmemowner", shmem_owner);
+	    if (!shmem_owner.isstring())
+		return false;
+
+	    // First do a pass to collect the final size
+	    SYS_SharedMemoryOutputStream os_count(NULL);
+	    {
+		openvdb::io::Stream vos(os_count);
+		geoSetVDBStreamCompression(vos, /*backwards_compatible*/false);
+		vos.write(grids, meta);
+	    }
+
+	    // Create the shmem segment
+	    UT_WorkBuffer shmem_key;
+	    shmem_key.sprintf("%s:%p", shmem_owner.buffer(), this);
+	    UT_SharedMemoryManager &shmgr = UT_SharedMemoryManager::get();
+	    SYS_SharedMemory *shmem = shmgr.get(shmem_key.buffer());
+	    if (shmem->size() != os_count.size())
+		shmem->reset(os_count.size());
+
+	    // Save the vdb stream to the shmem segment
+	    SYS_SharedMemoryOutputStream os_shm(shmem);
+	    {
+		openvdb::io::Stream vos(os_shm);
+		geoSetVDBStreamCompression(vos, /*backwards_compatible*/false);
+		vos.write(grids, meta);
+	    }
```

Comment:
I'm curious about all this new shared memory code and what it means in practice?

Reply:
This code has been sitting in Houdini for quite a few years now. They were added for performance so that we can stream the grids to memory for the Render View (IPR) pane which is shared with an external mantra process. As I wasn't involved in this change so I'm not sure what was the performance difference.

Reply:
Got it, thanks!